### ISSUE TITLE
postcss-custom-media v9 not replacing variables

### ISSUE THREAD

**Issue by gpoitch** [csstools/postcss-plugins#721]
### Bug description

Since updating to v9+ (9.0.1) postcss-custom-media is not compiling and replacing custom media variables. Have been using v8 for a while with no issue.

### Source CSS

```css
@custom-media --tabletUp (min-width: 768px);

@media (--tabletUp) {
  a { color: red; }
}
```

### Expected CSS

```css
@media (min-width: 768px) {
  a { color: red; }
}
```

### Actual CSS

```css
@media (--tabletUp) {
  a { color: red; }
}
```

### Extra config

Used via webpack:

```js
{
  loader: 'postcss-loader',
  options: {
    postcssOptions: {
      plugins: [postcssCustomMedia(), autoprefixer()]
    }
  }
}
```

### What plugin are you experiencing this issue on?

PostCSS Custom Media Queries

### Plugin version

9.0.1

### What OS are you experiencing this on?

macOS

### Node Version

19.1.0

### Validations

- [X] Follow our [Code of Conduct](https://github.com/csstools/postcss-plugins/blob/master/CODE_OF_CONDUCT.md)
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### Would you like to open a PR for this bug?

- [ ] I'm willing to open a PR

**Comment by romainmenke** [comment]
@gpoitch Thank you for reaching out!

I did a quick check and it works on our end : https://github.com/csstools/postcss-plugins/pull/722/files

Are there maybe other plugins that can be interfering?

**Comment by romainmenke** [comment]
Also see : https://github.com/csstools/postcss-plugins/issues/705#issuecomment-1316647241

It seems that for some users something is "sticky" and after updating the `@custom-media` are not transformed at first.

I don't know that user's or your setup.
Maybe there is a cached config or a watcher still running?

**Comment by gpoitch** [comment]
Same issue with postcss-custom-media as the only plugin and cleaning any possible caches and reinstalling all node_modules.

I didn't see a changelog but maybe something changed with how separate files are combined? My setup has all the variables in its own file.

```css
@custom-media --tabletUp (min-width: 768px);
@custom-media --desktopUp (min-width: 1025px);
--primaryColor: blue;
--textColor: black;
```

**Comment by romainmenke** [comment]
🤔 We did indeed change something could affect that.

We now use `prepare` to make it safe to re-use the same plugin instance multiple times.
https://github.com/csstools/postcss-plugins/blob/postcss-preset-env--v8/plugins/postcss-custom-media/src/index.ts#L25-L26

Can you share a minimal example of two CSS files?

- one file with the `@custom-media` declarations
- one file where you use them

I want to make sure I am accurately reproducing how you are bundling.

**Comment by gpoitch** [comment]
Here is a minimal example:
  
variables.css:
```css
@custom-media --tabletUp (min-width: 768px);

:root {
  --primaryColor: blue;
}

```


app.css:
```css
.box {
  width: 100px;
  height: 100px;
  background: var(--primaryColor);
}

@media (--tabletUp) {
  .box {
    width: 300px;
    height: 300px;
  }
}
```

app.js:
```js
import './variables.css'
import style from './app.css'

const box = document.createElement('div')
box.className = style.box
document.body.appendChild(box)
```

**Comment by romainmenke** [comment]
I can't get that to work, even with version 8.
Where there any other dependency updates?

**Comment by gpoitch** [comment]
Nothing else changed. Had been working fine for years on v8 and works correctly when I downgrade back to 8

**Comment by romainmenke** [comment]
Found it, and can confirm.

_It wasn't working for me because I was checking this in the same workspace as the updated plugins and WebPack loaders do not always find the right version in that case._

This is a very bad bug in the postcss loader for webpack :/
It appears to work in `v8`, but it doesn't.

It is one big race condition.

- your variables file is likely small and parses quickly
- your other files are likely larger and parse slower
- all files are handled in parallel
- in `v8` there is a shared object to store the `@custom-media` object
- variables finish first and object gets populated
- other files finish second and can use the `@custom-media` object

This is what happens when files are of similar length :
Each time the contents change I ran `npm run build`

https://user-images.githubusercontent.com/11521496/203084964-e5b04c2e-71df-4b38-96b7-420ed8a02cf1.mov

**Comment by romainmenke** [comment]
@gpoitch This issue is somewhat related : https://github.com/postcss/postcss-import/issues/304

> The reason I use postcss-import is that I also use postcss-cssnext and have global variables defined in a variables.css that I need to import in all of my css files. And because cssnext needs to work with these values like color modifications, the variables need to be accessible in that stage.

If you want to share state between files you could use `postcss-import`.

**Comment by romainmenke** [comment]
I am sure many other will run into this issue :/

Unfortunately I fear that we won't fix this unless something major is also changed in Webpack / PostCSS Loader.

So maybe better to adopt `postcss-import` as it fixes the root cause?

**Comment by gpoitch** [comment]
I'd rather not have to import in every file. My non-simplified setup has 100s of css files and then it's another step a team needs to know for new files.

I am familiar with the parallel nature from writing other webpack plugins. 

Perhaps there is a way to apply it to the compiled files instead of individually via a loader.

**Comment by gpoitch** [comment]
Just tried that it only performs the replacement in files with the `@import` so would have to do on every file

**Comment by romainmenke** [comment]
🤔 that works on my end.
In this example it is picking up on the `@import`'s done in `app.css` even if the imported CSS needs to be used in `styles-a.css`. `styles-a.css` doesn't have any `@import`'s

```css
/* app.css */
@import url(./media-queries.css);
@import url(./media-queries-2.css);
@import url(./styles-a.css);
```

```css
/* media-queries.css */
@custom-media --tablet-up (min-width: 768px);
```

```css
/* media-queries-2.css */
@custom-media --tablet-down (max-width: 768px);
```

```css
/* style-a.css */
:root {
	--c: hsl(50deg 50% 50%);
}

.a {
	color: var(--c);
}

@media (--tablet-up) {
	.a {
			color: pink;
		}
}

@media (--tablet-down) {
	.a {
		color: hotpink;
	}
}
```

```js
// index.js
import './app.css';
```

---

Result :

```css
:root {
	--c: hsl(50, 50%, 50%);
}
.a {
	color: hsl(50, 50%, 50%);
	color: var(--c);
}
@media (min-width: 768px) {
	.a {
			color: pink;
		}
}
@media (max-width: 768px) {
	.a {
		color: hotpink;
	}
}
```

**Comment by gpoitch** [comment]
I see. I think that's because you imported @import url(./styles-a.css); That means I would have to import every file by name which is worse than importing the variables into every file.

**Comment by romainmenke** [comment]
In that case a different Webpack plugin to get the desired effect might be better.

Given that the issue is the result of how Webpack / PostCSS Loader works and that this plugin in isolation is working as intended I will close this issue if that is ok?

**Comment by gpoitch** [comment]
Yea whatever you think. 
I see the `importFrom` config option I actually used at one point to get this to work but it worked without it (I guess accidentally) and that is now no longer supported anyway (still in docs fyi).

**Comment by gpoitch** [comment]
For future reference: using this is the solution: https://github.com/csstools/postcss-plugins/tree/main/plugins/postcss-global-data#readme
